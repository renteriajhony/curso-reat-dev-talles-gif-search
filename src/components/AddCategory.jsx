import { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({ onAddCategory }) => {
  const [inputValue, setInputValue] = useState('');

  const onInputChange = ({ target }) => {
    setInputValue(target.value);
  };

  const onSubmit = (event) => {
    event.preventDefault();
    const newValue = inputValue.trim();
    if (newValue.length <= 1) return;
    onAddCategory(newValue);
    setInputValue('');
  };
  return (
    <form onSubmit={onSubmit} aria-label='form'>
      <input
        type="text"
        placeholder="Search gifs..."
        value={inputValue}
        onChange={onInputChange}
      />
    </form>
  );
};

AddCategory.propTypes = {
  onAddCategory: PropTypes.func.isRequired,
};
