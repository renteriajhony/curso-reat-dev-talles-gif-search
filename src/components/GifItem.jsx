import PropTypes from 'prop-types';

export const GifItem = ({ title, url }) => {
  return (
    <div className="card">
      <img key={url} src={url} alt={title} width={150} height={300} />
      <p key={title}>{title}</p>
    </div>
  );
};

GifItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
};
