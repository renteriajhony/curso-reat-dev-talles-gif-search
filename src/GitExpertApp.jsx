import { useState } from 'react';
import { AddCategory, GifGrid } from './components';

export const GitExpertApp = () => {
  const [categories, setCategorie] = useState(['One Punch Man']);

  const onAddCategory = (category) => {
    categories.find(
      (item) => item.toLowerCase().trim() === category.toLowerCase().trim(),
    )
      ? alert('Category already exists!')
      : setCategorie([category, ...categories]);
  };

  return (
    <>
      <h1>GitExpertApp</h1>
      <AddCategory onAddCategory={onAddCategory} />

      {categories.map((category) => (
        <GifGrid key={category} category={category} />
      ))}
    </>
  );
};
