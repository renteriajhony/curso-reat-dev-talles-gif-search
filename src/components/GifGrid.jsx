import PropTypes from 'prop-types';
import { GifItem } from './GifItem';
import { useFetchGifs } from '../hooks/useFetchGifs';

export const GifGrid = ({ category }) => {
  const { gifs, isLoading } = useFetchGifs(category);

  return (
    <>
      <h3>{category}</h3>
      {isLoading && <p>Loading...</p>}
      <div className="card-grid" key={category}>
        {gifs.map((gif) => (
          <GifItem key={gif.id} {...gif} />
        ))}
      </div>
    </>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired,
};
