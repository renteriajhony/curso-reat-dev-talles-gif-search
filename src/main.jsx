import React from 'react';
import ReactDOM from 'react-dom/client';
import { GitExpertApp } from './GitExpertApp.jsx';
import './styles.css';

ReactDOM.createRoot(document.getElementById('root')).render(
  //TODO: para producción, usar React.StrictMode, que se encarga de detectar errores de diseño, y manejarlos de manera segura
  // <React.StrictMode>
  //   <GitExpertApp />
  // </React.StrictMode>,

  <GitExpertApp />,
);
