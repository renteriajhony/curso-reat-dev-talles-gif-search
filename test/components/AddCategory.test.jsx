import { render, screen, fireEvent } from '@testing-library/react';
import { AddCategory } from '../../src/components';

describe('Pruebas en <AddCategory/>', () => {
  const onAdd = jest.fn();

  test('debe hacer match con snapshop ', () => {
    const { container } = render(<AddCategory onAddCategory={onAdd} />);
    expect(container).toMatchSnapshot();
  });

  test('debe cambiar el valor de la caja de texto ', () => {
    render(<AddCategory onAddCategory={onAdd} />);
    const input = screen.getByPlaceholderText('Search gifs...');
    fireEvent.change(input, { target: { value: 'Goku' } });
    expect(input.value).toBe('Goku');
  });

  test('debe llamar onSubmit, si el input tiene un valor valido ', () => {
    const inputValue = 'Goku';
    render(<AddCategory onAddCategory={onAdd} />);
    const input = screen.getByRole('textbox');
    const form = screen.getByRole('form');
    fireEvent.change(input, { target: { value: inputValue } });
    expect(input.value).toBe(inputValue);
    fireEvent.submit(form);
    expect(input.value).toBe('');
    expect(onAdd).toHaveBeenCalled();
    expect(onAdd).toHaveBeenCalledWith(inputValue);
    expect(onAdd).toHaveBeenCalledTimes(1);
  });

  test('No dene llamar onSubmit si el input no tiene un valor valido ', () => {
    const inputValue = 'G';
    const onAdd2 = jest.fn();
    render(<AddCategory onAddCategory={onAdd} />);
    const input = screen.getByRole('textbox');
    const form = screen.getByRole('form');
    fireEvent.change(input, { target: { value: inputValue } });
    expect(input.value).toBe(inputValue);
    fireEvent.submit(form);
    expect(onAdd2).not.toHaveBeenCalled();
    expect(onAdd2).toHaveBeenCalledTimes(0);
  });
});
