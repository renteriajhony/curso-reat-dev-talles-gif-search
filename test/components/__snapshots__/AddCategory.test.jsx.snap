// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddCategory/> debe hacer match con snapshop  1`] = `
<div>
  <form
    aria-label="form"
  >
    <input
      placeholder="Search gifs..."
      type="text"
      value=""
    />
  </form>
</div>
`;
